crates/holochain/src/conductor/conductor.rs

```diff
@@ -3028,16 +3054,64 @@ mod misc_impls {
         /// Inject records into a source chain for a cell.
         /// If the records form a chain segment that can be "grafted" onto the existing chain, it will be.
         /// Otherwise, a new chain will be formed using the specified records.
+        //
+        // TODO: could integrate this better with [`SourceChain::flush`] when re-visit our workflows
+        //       and use them more consistently for moving data around
         pub async fn graft_records_onto_source_chain(
```

Comment:
Perhaps it's more appropriate to just hold onto a DB transaction for the duration of this function, rather than formally locking the chain, since if there's an error, it'll be a pain to unlock it. That will prevent concurrent writes. wyt?

Reply:
That could work. I added the ability to claim the write semaphore for longer than just a `write_async` in the countersigning PR. That was used to allow a read, followed by the CHC push, followed by a write, without any writes being able to get mixed in. The same thing should work here.

Reply:
I wound up writing this to just keep a db transaction open the whole time

Reply:
BTW, I forgot to mention that I ran into trouble with this. The long-lived semaphore let me create a deadlock that I had a hard time finding. If I had been aware of the new technique you had added maybe I would have known better where to look. so maybe this isn't as much of a problem for future devs.

Also, when I finally found it and made use to re-use the existing semaphore, it was pretty ugly, since any function that wants to use it has to return ownership when it's done. I don't think this is a great pattern going forward, for the reasons I mentioned, as well as it opening the possibility of async functions being called while the semaphore is held, potentially blocking the database for unacceptably long periods of time. Making async calls during a transaction may be necessary in some cases (it was for me to integrate CHC sync with your countersigning work) but it would be good to discourage this as much as possible.

I would prefer to find ways to design things so that atomic blocks are truly short and synchronous, and if that's not possible, to find ways to break them up into multiple transactions without breaking data integrity.

Since we're not stabilizing countersigning or CHC for now though, maybe this is fine?